src/buffer.rs

```diff
@@ -406,54 +395,62 @@ where Container: ArrayLike<T>, T: Primitive + 'static, PixelType: Pixel<T> + 'st
         (0, 0, self.width, self.height)
     }
 
-    fn get_pixel(&self, x: u32, y: u32) -> PixelType {
+    fn get_pixel(&self, x: u32, y: u32) -> P {
         *self.get_pixel(x, y)
     }
 
-    fn get_pixel_mut(&mut self, x: u32, y: u32) -> &mut PixelType {
+    fn get_pixel_mut(&mut self, x: u32, y: u32) -> &mut P {
         self.get_pixel_mut(x, y)
     }
 
-    fn put_pixel(&mut self, x: u32, y: u32, pixel: PixelType) {
+    fn put_pixel(&mut self, x: u32, y: u32, pixel: P) {
         *self.get_pixel_mut(x, y) = pixel
     }
 
     /// Put a pixel at location (x, y), taking into account alpha channels
     #[deprecated = "This method will be removed. Blend the pixel directly instead."]
-    fn blend_pixel(&mut self, x: u32, y: u32, p: PixelType) {
+    fn blend_pixel(&mut self, x: u32, y: u32, p: P) {
         self.get_pixel_mut(x, y).blend(&p)
     }
 }
 
-impl<Container, T, PixelType> Index<(u32, u32)>
-for ImageBuffer<Container, T, PixelType>
-where Container: ArrayLike<T>, T: Primitive + 'static, PixelType: Pixel<T> + 'static {
-    type Output = PixelType;
-    fn index(&self, &(x, y): &(u32, u32)) -> &PixelType {
+impl<P: Pixel + 'static, Container: ArrayLike<P::Subpixel>> Index<(u32, u32)>
+    for ImageBuffer<P, Container> where P::Subpixel: 'static {
+
+    type Output = P;
+
+    fn index(&self, &(x, y): &(u32, u32)) -> &P {
         self.get_pixel(x, y)
     }
 }
- 
+
 // concrete implementation for `Vec`-baked buffers
-impl<T, PixelType> ImageBuffer<Vec<T>, T, PixelType>
-where T: Primitive + 'static, PixelType: Pixel<T> + 'static {
-    /// Creates a new image buffer based on a `Vec<T>`.
-    pub fn new(width: u32, height: u32) -> ImageBuffer<Vec<T>, T, PixelType> {
+// TODO: I think that rustc does not "see" this impl any more: the impl with
+// Container meets the same requirements. At least, I got compile errors that
+// there is no such function as `into_vec`, whereas `into_raw` did work, and
+// `into_vec` is redundant anyway, because `into_raw` will give you the vector,
+// and it is more generic.
```

Comment:
I'm a bit worried about that. Is this a compiler bug or something I did wrong? Does `rustc` still find `new`? If not this is bad. How is this supposed to work? Afaik one can’t just add a generic constructor to the ArrayLike trait because static methods are not object safe. 

PS: I just added `into_vec`  etc. to match the stdlib.